### ISSUE TITLE
Creating single release with multiple build artifacts on Github

### ISSUE THREAD

**Issue by Flamefire** [travis-ci/travis-ci#10230]
As mentioned in #8053 especially for C++ you have multiple builds running each producing an artifact for e.g. windows, linux and osx. Those shall be deployed as 1 release (e.g.) on Github.

Currently this does not seem to be possible.

Description copied:

I'm building a linux and osx binary and have setup to automatically add a release for them on github. I've setup the name of the release to be the tag, so it is the same for both the linux and osx build.
Still travis is adding two releases with the same tag/name instead of putting both binaries on the same release.   
I also build on appveyor for the windows binary. That works fine (it either creates a new or adds to the existing release with the same name).

Goal: Run (at least) 2 different jobs on travis (e.g. linux and OSX as the image) and deploy the artifacts to a single Github release. This is for an OpenSource project so a paid AWS S3 instance is out of question.

The Github API supports editing releases and adding files to it, so it should not be a problem in general.

Is there any full tutorial/example travis.yml to do this?

**Comment by BanzaiMan** [comment]
This is certainly possible now.

The general logic would be like this:

1. Configure jobs that generate artifacts you want in a release. For each of these, define a GitHub Releases deployment with `draft: true`.
1. Choose one of these jobs as the "publisher" job, and set it aside as the sole job in a different Build Stage (Publish, say), which will run at the end of the build. For this job, you would set `draft: false`, so that the resulting Release will be publicly visible.


```yaml
deploy: &deploy_base
  provider: releases
  api_key: $GITHUB_TOKEN
  draft: true
  skip_cleanup: true
  file: file-$TRAVIS_OS_NAME.txt
  on:
    all_branches: true

install: skip
script: date > file-$TRAVIS_OS_NAME.txt

jobs:
  include:
    - stage: compile macos assets
      os: osx
      deploy: 
        <<: *deploy_base
    - stage: compile linux assets and publish
      deploy:
        <<: *deploy_base
        draft: false
```

**Comment by Flamefire** [comment]
wouldn't the top-level deploy result in another job doing an additional deployment?

I know that e.g. gitlab supports "hidden keys" prefixed with dot which are ignored by the CI bot but can be referenced for e.g. this use case.

**Comment by BanzaiMan** [comment]
@Flamefire "It depends." I've updated the snippet above to remove that ambiguity.

**Comment by Flamefire** [comment]
Ok thank you. Only remaining thing I'm now concerned of is the following from the docu:

> Notice that this tag is immediately available on GitHub, and thus will trigger a new Travis CI build, unless it is prevented by other means.

To create a release you have to have a tag. We want to do "nighly" releases on every master push/merge. If you create/change a tag manually (`git tag $TRAVIS_TAG` as per docu) or none at all (which creates a tag "untagged-*") another build is triggered.

The linked "other means" does not state HOW to avoid this duplicate build of the same commit (just with the new tag)

**Comment by BanzaiMan** [comment]
@Flamefire You are right about the documentation link. A more appropriate place is https://docs.travis-ci.com/user/customizing-the-build/#safelisting-or-blocklisting-branches, which explains how to exclude branches and tags from getting built.

**Comment by Flamefire** [comment]
Ok thanks for the clarification on that. Am I right in assuming that manually adding the tag in the `pre_deploy` step via `git tag $TRAVIS_TAG` also results in a build being started? Does this build also get started, if the tag is not changed when e.g. set with `get tag -f $TRAVIS_TAG`? How could I avoid this?

**Comment by BanzaiMan** [comment]
> Am I right in assuming that manually adding the tag in the `pre_deploy` step via `git tag $TRAVIS_TAG` also results in a build being started?

No. Unless you push this tag to GitHub, this tag will not be visible to GitHub (or Travis CI, by extension).

**Comment by Flamefire** [comment]
Oh so I don't need to push that tag and travis + (more importantly) Github will create the release without a tag on the remote? Didn't know this and didn't expect this as the Github API needs a tag if I'm not mistaken.

**Comment by BanzaiMan** [comment]
> Github will create the release without a tag on the remote?

No. That is incorrect. When a Release is created and the tag is not specified, GitHub will create a *tag* and associate that tag to the release that we've just created.

> the Github API needs a tag if I'm not mistaken.

This is true (docs: https://developer.github.com/v3/repos/releases/#create-a-release). But we use Octokit, and this is the behavior we have observed.

**Comment by Flamefire** [comment]
> When a Release is created and the tag is not specified, GitHub will create a tag and associate that tag to the release that we've just created.

And this tag is not "pushed" in a way that travis reacts on, I guess

**Comment by BanzaiMan** [comment]
🤔 To think, it might be Octokit that's adding `untagged-*`.

**Comment by BanzaiMan** [comment]
I don't see `untagged` in Octokit code base, so I suspect it's GitHub. See also: https://github.com/octokit/octokit.rb/commit/0619c34c253d2a47204f21a74daccbf79ae0b6f6.

**Comment by beemaninsuranceagents** [comment]
install: skip
script: date > file-$TRAVIS_OS_NAME.txt

jobs:
  include:
    - stage: compile macos assets
      os: osx
      deploy: &deploy_base
        provider: releases
        api_key: $GITHUB_TOKEN
        draft: true
        skip_cleanup: true
        file: file-$TRAVIS_OS_NAME.txt
        on:
          all_branches: true
    - stage: compile linux assets and publish
      deploy:
        <<: *deploy_base
        draft: false

**Comment by daniellivingston** [comment]
@BanzaiMan , running the above script exactly as you provided results in these releases:

    untagged-1a57f876e5693d27bd45
    file-linux.txt

    Draft
    file-osx.txt

In other words, still two separate releases, with one being a draft.
Is there no way to add `file-linux.txt` to the `file-osx.txt` draft && publish?

**Comment by grntmhqo** [comment]
### 
    - add_file-linux.txt >(file-osx.txt)
    - draft && publish?

**Comment by daniellivingston** [comment]
@grntmhqo I'm not sure what you mean.

**Comment by BanzaiMan** [comment]
@daniellivingston Not sure how your build is configured differently, but you can see mine at https://github.com/BanzaiMan/travis_production_test/blob/multi-os-release-3/.travis.yml

This tag just ran [this build](https://travis-ci.org/BanzaiMan/travis_production_test/builds/479983050) and created [this release](https://github.com/BanzaiMan/travis_production_test/releases/tag/multi-os-release-3) without any "untagged-*" release.

**Comment by BanzaiMan** [comment]
See [this commit](https://github.com/BanzaiMan/travis_production_test/commit/77c93be5ed5efbc16029e66958222484d44ec99f) for the one using YAML anchor and alias.

Build: https://travis-ci.org/BanzaiMan/travis_production_test/builds/479988575
Release: https://github.com/BanzaiMan/travis_production_test/releases/tag/multi-os-release-4

**Comment by daniellivingston** [comment]
@BanzaiMan Brilliant! The non-aliased version works beautifully for me. I must have configured the YAML anchor + alias version incorrectly.

**Comment by hallowf** [comment]
I have tried both the non aliased version and the alias one and i'm still getting 2 releases however i'm compiling for windows and linux

My Config:
```
jobs:
  include:
    - stage: Linux build
      os: linux
      language: python
      sudo: required
      dist: xenial
      python: '3.7'
      env:
      - PY=python3
      - UPX_PATH="$PWD"/UPX
      deploy:
        provider: releases
        api_key:
        skip_cleanup: true
        file_glob: true
        file: dist/*
        draft: true
        on:
          repo: hallowf/py-backs-cli
    - stage: Windows build
      os: windows
      language: shell
      python: 3.7
      env:
      - PY=py
      - UPX_PATH="$PWD"/UPX
      deploy:
        provider: releases
        api_key:
        skip_cleanup: true
        file_glob: true
        file: dist/*
        draft: true
        on:
          repo: hallowf/py-backs-cli
```

Results in 2 different releases each with it's own binary

**Comment by kvnndr0** [comment]
Banzaiman

**Comment by mmcc007** [comment]
This is how I was able to publish 2 artifacts generated by a single job.

The 2 artifacts are saved in the cache by the first job. 

One artifact is published by the first job as a draft. The second artifacts is then found in the cache and published in a second job.

You can see the jobs running here:
https://travis-ci.com/mmcc007/screenshots/builds/103481608
and the artifacts published here:
https://github.com/mmcc007/screenshots/releases/tag/0.0.14

Seems like a separate job is required to publish each additional artifact. If there is a better way to publish multiple artifacts generated by a single job (when publishing to the same place), I am not aware (I'd like to know).

Seems like a single job should be able to publish multiple artifacts to a single destination.

**Comment by polarathene** [comment]
@BanzaiMan Your example works for me(creates a draft, then updates the file and publishes from the draft), the tag triggered another build but that can be avoided, non-issue.

What your example doesn't allow for is updating a draft release and keeping it as a draft release instead of publishing a release. Try your approach with a third build/file involved. **Drafts don't get updated** like tagged release/pre-release seems to, **instead a new draft is created**.. 

I'm not quite sure how the 2nd deploy manages to update and release the prior draft, **is it a bug** related to the `draft: true` logic creating a new draft instead?

In the github releases web GUI, you can see every time you click "Save Draft" button the URL updates with a new `untagged-` url. This is a tag that will be used(created upon publish) if one is not specified when publishing the release(**nothing to do with octokit generating it**). If you want to update the draft, you'll need to pass that specific tag(well full URL that includes it) to the [octokit edit release URL param](https://octokit.github.io/octokit.rb/Octokit/Client/Releases.html#update_release-instance_method)( it's not mentioned [here](https://developer.github.com/v3/repos/releases/#get-the-latest-release), presumably it's equivalent to `release_id` in that docs link?). Problem is, that updates the random tag used to reference it afaik. 

How does Travis know when to create a new draft or update an existing one? It seems like it's only capable of creating new draft releases without anyway to reference a draft to update(in existing build or prior build deploy). Perhaps you could have "update_on: tag_name/tag_commitish/name/etc", so if a draft release exists it can be updated based on what's provided there, otherwise create a new one?

@mmcc007 You can upload multiple files via array or glob patterns:

```YAML
  file_glob: true
  file:
    - dist/**/*.exe
    - dist/**/*.dmg
```

Or did I misunderstand you?

**Comment by mmcc007** [comment]
@polarathene the file array approach worked for me for one job with multiple artifacts...
https://travis-ci.com/mmcc007/screenshots/builds/104898262
https://github.com/mmcc007/screenshots/releases/tag/0.0.15
Thnx!

**Comment by polarathene** [comment]
@BanzaiMan I came across the source related to this from one of your comments in a similar issue.

The URL param I referred to that was important to draft releases isn't documented in Travis docs and is an available option but different name than the two octokit docs I linked:

[releases.rb#L104](https://github.com/travis-ci/dpl/blob/c53b1bcda2ec62545bfd85df658311c0508845ad/lib/dpl/provider/releases.rb#L104):

```Ruby
release_url = "https://api.github.com/repos/" + slug + "/releases/" + options[:release_number]
```

If that param/option wasn't provided, then it iterates through the `releases` results to check for any tags(tags on that commit or value of `$TRAVIS_TAG`) that match a release's `tag_name` value...

So that works alright if the deploy is set only to deploy `on: tags: true` and a commit is tagged to trigger a build. However, some users want a workflow where a draft release is updated with each commit and then making the release that actually assigns the release tag to the commit they want to release on. 

Probably makes sense to make a release with a tag, then publish it which triggers a Travis build / deploy to upload artifacts, or to go the other route and tag the release commit, then have Travis deploy a draft release if the user wants to fill out details once that's available then publish. 

But I and other users seem to have been tripped up a fair bit with this process(docs really haven't been clear for how to go about it, especially with multiple OS mixed with draft adding to confusion of the separate releases and various advice/solutions/workarounds people have been sharing over the years. 

While trying to figure out and solve the issue, I was triggering builds without tags and getting very confused why I couldn't get a single draft release with the bundles(earlier advice elsewhere suggested sending artifacts to AWS S3 or similar in order to achieve this, and/or that since deploys were tied to their related job, they couldn't deploy to the same release, but now it's clear you can target and update a specific release).

---

**TL;DR:**

Presently, you can workaround this issue(multiple build artifacts to a single draft release with no tagged commit) by setting the value of `STRAVIS_TAG` to be/match the value of a draft release with the same `tag_name`(what is compared to tags for the given commit or with the value of `$TRAVIS_TAG` when it has a value).

Here is a minimal working example:

```YAML
if: tag IS blank

language: ruby
  
matrix:
  include:
    - name: "macOS Build"
      os: osx
      
    - name: "Linux Build"
      os: linux

script: date > file-$TRAVIS_OS_NAME.txt

before_deploy:
  - export TRAVIS_TAG="preview"

deploy:
  provider: releases
  api_key: $GH_TOKEN
  skip_cleanup: true
  draft: true
  overwrite: true
  file_glob: true
  file: file-*.txt
```

### Details

It will now build on commits(while ignoring tags being added such as when publishing without existing tag) and if the value of `$TRAVIS_TAG`("preview" in this case) doesn't exist as a tag_name in any draft release, it will create the draft release. Both builds will send files to that same draft release.

You can then publish the release. You should change the `tag_name`  field manually if publishing this way, otherwise the next build will override the previous files(`overwrite: true` if filenames are the same, otherwise appends/adds to the release) and change the release back to draft. If you don't pubilsh, then each new commit will update the draft release.

If you gave a different tag when manually publishing, your next commit will build a new draft release tagged "preview" like before until you're ready to release. Do note that if you have other draft releases with the same `tag_name` value(shouldn't really happen), then only one of these entries/releases gets updated by Travis(oldest one afaik). I only encountered that by seeing what happens if there are two draft releases with the same `tag_name`. 

Blank draft release `tag_name`s(Empty string value: "") will not match an existing draft release, Travis instead will create a new release(which publishes the `untagged` value in the URL of release edit page, as a tag).

Alternatively, you could only deploy when you add a tag to a commit(or publish a release that creates the tag):

```YAML
language: ruby
  
matrix:
  include:
    - name: "macOS Build"
      os: osx
      
    - name: "Linux Build"
      os: linux

script: date > file-$TRAVIS_OS_NAME.txt

deploy:
  provider: releases
  api_key: $GH_TOKEN
  skip_cleanup: true
  file: file-$TRAVIS_OS_NAME.txt
  on:
    tags: true
```

Not much reason for having `overwrite: true` in this case, you could add `draft: true` if you want to make any additions to the release manually or control when it's actually published, because the commit is tagged, you don't need the `before_deploy` workaround, the tag for the commit will be used for the release, or if you have an existing draft release with the tag value in the ``tag_name` field, it'll upload any artifacts to that and release(if you don't specify `draft: true`), else wait for you to manually publish.

This approach will incur double builds though(once on commit, and then once again upon tagging. If that's an issue you can cancel one of the builds. If the commit is related to version bump for the release, there's probably a way to filter out by that commit message via some regex + `tag IS blank` condition to exclude commits like that from builds. (There's some options for skipping builds by detecting keywords in the commit message already, but afaik that'd also skip regardless of a tag being present).

**Comment by soywod** [comment]
Thanks to this issue, I was able to also set up cross-building and deploying exes to the same release. I noticed that the option `deploy on tags: true` is the key. When jobs point to the same tag, builds are added to the same release. Here [the release](https://github.com/unfog-io/unfog-cli/releases/tag/v0.1.2), [the build](https://travis-ci.org/unfog-io/unfog-cli/builds/617108387) and [the travis conf](https://github.com/unfog-io/unfog-cli/blob/0208de9a1a16d50cf744add6a2474fd1a6974dd6/.travis.yml).

**Comment by MarkCallow** [comment]
I had this working using dpl v1 without any of the shenanigans discussed here:

```yaml
deploy:
  - provider: releases
    # edge: true # Use bleeding edge (dplv2)
    api_key: $GITHUB_TOKEN  # Set in the repo settings page as a secure variable
    file_glob: true
    file:
      - build-macos/KTX-Software-*.pkg
    skip_cleanup: true
    draft: true
    prerelease: true
    on:
      tags: true
      condition: $TRAVIS_OS_NAME = osx
  - provider: releases
    #edge: true # Use bleeding edge (dplv2)
    api_key: $GITHUB_TOKEN  # Set in the repo settings page as a secure variable
    file_glob: true
    file:
      - build-linux-release/KTX-Software-*-Linux.*
      - build-web-release/KTX-Software-*-Web-*.zip
    skip_cleanup: true
    draft: true
    prerelease: true
    on:
      tags: true
      condition: $TRAVIS_OS_NAME = linux
```

If I uncomment the `edge: true` lines then it breaks and I get 2 draft releases with the same name. Is this issue about dpl v1 or dpl v2?

Why does Travis make this so damn difficult? I'm sure there must be many people building for both Linux and macOS who want to put the assets in a single release.
